src/core/string.h

```diff
@@ -0,0 +1,231 @@
+// Copyright (c) 2014-2015 Dropbox, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef PYSTON_CORE_STRING_H
+#define PYSTON_CORE_STRING_H
+
+#include <stdexcept>
+
+namespace pyston {
+
+class String {
+public:
+    explicit String(const std::string& s, char* _ptr = NULL) : ptr(_ptr), own(_ptr == NULL) {
```

Comment:
Can we get a comment about the ownership strategy?  I was somewhat confused by the way this works, since I was assuming it would work similarly to llvm::StringRef in the non-owning case (ie would just reference the original data rather than copy it).